src/sql/base/browser/ui/table/treeGrid.ts

```diff
@@ -0,0 +1,177 @@
+					if (data.cell === 0) {
+						this.setCellExpandedState(data.row, this.getExpandableColumnIndex(), false);
+					}
+				} else if (event.keyCode === KeyCode.RightArrow) {
+					// Right arrow on last cell of the collapsed row expands it.
+					if (data.cell === (this._grid.getColumns().length - 1)) {
+						this.setCellExpandedState(data.row, this.getExpandableColumnIndex(), true);
+					}
+				}
+			}
+			return false;
+		});
+
+		this._grid.onRendered.subscribe((e, data) => {
+			const visibleRows = this._grid.getViewport();
+			for (let i = visibleRows.top; i < visibleRows.bottom; i++) {
+				const rowData = this._data.getItem(i);
+				// Getting the row div that corresponds to the data row
+				const rowElement = this._tableContainer.querySelector(`div [role="row"][aria-rowindex="${(i + 1)}"]`);
+				// If the row element is found in the dom, we are setting the required aria attributes for it.
+				if (rowElement) {
+					if (rowData.expanded !== undefined) {
+						rowElement.ariaExpanded = rowData.expanded;
+					}
+					if (rowData.level !== undefined) {
+						rowElement.ariaLevel = rowData.level;
+					}
+					if (rowData.setSize !== undefined) {
+						rowElement.ariaSetSize = rowData.setSize;
+					}
+					if (rowData.posInSet !== undefined) {
+						rowElement.ariaPosInSet = rowData.posInSet;
+					}
+				}
+			}
+			return false;
+		});
+	}
+
+	override setData(data: Array<T>): void;
+	override setData(data: TableDataView<T>): void;
+	override setData(data: AsyncDataProvider<T>): void;
+	override setData(data: Array<T> | TableDataView<T> | AsyncDataProvider<T>): void {
+		if (data instanceof TableDataView || data instanceof AsyncDataProvider) {
+			this._data = data;
+		} else {
+			this._data = new TableDataView<T>(data, undefined, undefined, defaultTreeGridFilter);
+		}
+		this.transformData(this._data);
+		this._grid.setData(this._data, true);
+		this._data.filter(this._grid.getColumns());
+	}
+
+	/**
+	 * This functions toggles the expanded state of a parent cell.
+	 * @param row row index of the parent cell
+	 * @param cell cell/ column index of the parent cell
+	 * @param expanded force a state on the parent cell
+	 */
+	private setCellExpandedState(row: number, cell: number, expanded?: boolean): void {
+		const rowData = this._data.getItem(row);
+		if (rowData['isParent'] && this._grid.getColumns()[cell].formatter === treeGridExpandableColumnFormatter) {
+			if (expanded === undefined) {
+				(<any>rowData).expanded = !rowData.expanded;
+			} else {
+				(<any>rowData).expanded = expanded;
+			}
+			this._data.filter(this._grid.getColumns());
+			this.rerenderGrid();
+			this.focus();
+		}
+	}
+
+	/**
+	 * Gets the index of the column that has expandable column formatter. This column contains the chevron
+	 * icon that indicates if the row is expanded or collapsed.
+	 */
+	private getExpandableColumnIndex(): number {
+		return this._grid.getColumns().findIndex(c => c.formatter === treeGridExpandableColumnFormatter);
```

Comment:
Something like this would always be the first column, right? Or at least I'm not sure of a scenario where it wouldn't be that - so for ease of implementation I'd suggest just making this only support having the first column. 

And overall I'm not sure I like the pattern of having the formatter for that column need to be defined by the user directly and always have to be treeGridExpandableColumnFormatter. This is a tree grid so we should be assuming it's always going to have this formatter. So that should be something that this class handles - not something the user does. 

I'd suggest having the user specify a column formatter as normal and then in this class when the columns are set you create a treeGridExpandableColumnFormatter that has a parameter for the "base" formatter (defaulting to text cell if none specified) and overwrite the first column to have this new formatter. Then when the formatter is called it calls the base formatter to get the "inner" content and wraps that in the div like you have above. (this is essentially the decorator pattern)

Reply:
Where?

Honestly though I'd try to avoid over-engineering this and just make it as simple as we can right now. Always having the first column be the expandable one is a lot simpler and lets us focus on making sure that everything works in that scenario and then later on if needed we can modify it to support other columns. 

(for example - if you have other columns then you have to consider what if someone sets expandable to true on multiple columns?)